prow/cmd/exporter/collector.go

```diff
@@ -0,0 +1,122 @@
+	// metrics as their label sets are dynamic. This is a take-our-own-risk action and also a
+	// compromise for implementing a metric with both dynamic keys and dynamic values in
+	// the label set.
+	// https://godoc.org/github.com/prometheus/client_golang/prometheus#hdr-Custom_Collectors_and_constant_Metrics
+}
+
+func (pjc prowJobCollector) Collect(ch chan<- prometheus.Metric) {
+	logrus.Debug("ProwJobCollector collecting ...")
+	prowJobs, err := pjc.lister.List(pjc.selector)
+	if err != nil {
+		logrus.WithError(err).Errorf("Failed to list prow jobs with selector '%s'", pjc.selector)
+		return
+	}
+	for _, pj := range prowJobs {
+		agent := string(pj.Spec.Agent)
+		pjLabelKeys, pjLabelValues := kubeLabelsToPrometheusLabels(pj.Labels, "label_")
+		pjLabelKeys = append([]string{"prow_job_name", "prow_job_namespace", "prow_job_agent"}, pjLabelKeys...)
+		pjLabelValues = append([]string{pj.Name, pj.Namespace, agent}, pjLabelValues...)
+		labelDesc := prometheus.NewDesc(
+			"prow_job_labels",
+			"Kubernetes labels converted to Prometheus labels.",
+			pjLabelKeys, nil,
+		)
+		ch <- prometheus.MustNewConstMetric(
+			labelDesc,
+			prometheus.GaugeValue,
+			// always 1 since there is only 1 prow job for each namespace and each prow job name
+			float64(1),
+			pjLabelValues...,
+		)
+		pjAnnotationKeys, pjAnnotationValues := kubeLabelsToPrometheusLabels(pj.Annotations, "annotation_")
+		pjAnnotationKeys = append([]string{"prow_job_name", "prow_job_namespace", "prow_job_agent"}, pjAnnotationKeys...)
+		pjAnnotationValues = append([]string{pj.Name, pj.Namespace, agent}, pjAnnotationValues...)
+		annotationDesc := prometheus.NewDesc(
+			"prow_job_annotations",
+			"Kubernetes annotations converted to Prometheus labels.",
+			pjAnnotationKeys, nil,
+		)
+		ch <- prometheus.MustNewConstMetric(
+			annotationDesc,
+			prometheus.GaugeValue,
+			float64(1),
+			pjAnnotationValues...,
+		)
+	}
+}
+
+var (
+	invalidLabelCharRE    = regexp.MustCompile(`[^a-zA-Z0-9_]`)
+	escapeWithDoubleQuote = strings.NewReplacer("\\", `\\`, "\n", `\n`, "\"", `\"`)
+)
+
+// https://github.com/kubernetes/kube-state-metrics/blob/1d69c1e637564aec4591b5b03522fa8b5fca6597/internal/store/utils.go#L60
+func kubeLabelsToPrometheusLabels(labels map[string]string, prefix string) ([]string, []string) {
+	labelKeys := make([]string, 0, len(labels))
+	for k := range labels {
+		labelKeys = append(labelKeys, k)
+	}
+	sort.Strings(labelKeys)
+
+	labelValues := make([]string, 0, len(labels))
+	for i, k := range labelKeys {
+		labelKeys[i] = fmt.Sprintf("%s%s", prefix, sanitizeLabelName(k))
+		labelValues = append(labelValues, escapeString(labels[k]))
+	}
+	return labelKeys, labelValues
+}
+
+func sanitizeLabelName(s string) string {
+	return invalidLabelCharRE.ReplaceAllString(s, "_")
+}
+
+// https://github.com/kubernetes/kube-state-metrics/blob/1d69c1e637564aec4591b5b03522fa8b5fca6597/pkg/metric/metric.go#L96
```

Comment:
I must admit I do not find these link comments very helpful without any context.

Reply:
Let me add the context (will revise the PR too).
The link reminds us of why we are doing this: because `kube-state-metrics` does it for the labels from pod.

Reply:
@petr-muller 
Please check if the added comments are helping.